rust_src/src/time.rs

```diff
@@ -18,6 +20,105 @@ use crate::{
 
 const LO_TIME_BITS: i32 = 16;
 
+pub type LispTime = lisp_time;
+
+impl LispTime {
+    pub fn into_vec(self, nelem: usize) -> Vec<EmacsInt> {
+        let mut v = Vec::with_capacity(nelem);
+
+        if nelem >= 2 {
+            v.push(self.hi);
+            v.push(self.lo.into());
+        }
+        if nelem >= 3 {
+            v.push(self.us.into());
+        }
+        if nelem > 3 {
+            v.push(self.ps.into());
+        }
+
+        v
+    }
+}
+
+impl PartialEq for LispTime {
+    fn eq(&self, other: &LispTime) -> bool {
+        self.hi == other.hi && self.lo == other.lo && self.us == other.us && self.ps == other.ps
+    }
+}
+
+impl Eq for LispTime {}
+
+impl PartialOrd for LispTime {
+    fn partial_cmp(&self, other: &LispTime) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for LispTime {
+    fn cmp(&self, other: &LispTime) -> Ordering {
+        self.hi
+            .cmp(&other.hi)
+            .then_with(|| self.lo.cmp(&other.lo))
+            .then_with(|| self.us.cmp(&other.us))
+            .then_with(|| self.ps.cmp(&other.ps))
+    }
+}
+
+#[allow(clippy::suspicious_arithmetic_impl)]
+impl Add for LispTime {
+    type Output = LispTime;
+
+    fn add(self, other: LispTime) -> LispTime {
+        let mut hi = self.hi + other.hi;
+        let mut lo = self.lo + other.lo;
+        let mut us = self.us + other.us;
+        let mut ps = self.ps + other.ps;
```

Comment:
These mutating, procedural calculations could be rewritten to be pure. My guess is that the compiler would be able to do more with it that way. I also think the logic would be clearer, but that's a matter of opinion.

Reply:
I am not so certain there is a pay off there.

```
let (us,ps) = if ps > 1_000_000 {
    (us - 1_000_000, ps + 1)
else {
    (us, ps)
};
let (us, lo) = if us > 1_000_000 {
    (us - 1_000_000, lo + 1)
} else {
    (us, lo)
};
let (lo, hi) = if lo > 1 << LO_TIME_BITS {
   (lo - (1 << LO_TIME_BITS), hi + 1)
} else {
    (lo, hi)
};
```

Is that really better?

Reply:
I would calculate auxiliary variables `ps_diff`, `us_diff` etc first, and then calculate `ps`, `us` etc just once, since I don't like variable shadowing (it's no better than mutation from the reader's perspective).

I also tend to think that using a different implementation is per se a good thing (as long as the different implementation is at least as fast, clear, etc).